python_modules/dagster/dagster/_core/definitions/solid_invocation.py

```diff
@@ -286,10 +286,14 @@ def type_check_gen(gen):
                     and output_def.is_required
                     and not output_def.is_dynamic
                 ):
-                    raise DagsterInvariantViolationError(
-                        f"Invocation of {solid_def.node_type_str} '{context.alias}' did not return"
-                        f" an output for non-optional output '{output_def.name}'"
-                    )
+                    if output_def.dagster_type.is_nothing:
+                        # implicitly yield None as we do in execute_step
+                        yield Output(output_name=output_def.name, value=None)
```

Comment:
https://github.com/dagster-io/dagster/blame/master/python_modules/dagster/dagster/_core/execution/plan/execute_step.py#L167-L171

we should really figure out how to refactor all this - at least the increasing test coverage should make that a confident refactor

Reply:
ya the branching code paths aspect is a continuing point of pain (my own sin)